src/main/java/com/example/online_shop_api/Service/UserService.java

```diff
@@ -218,4 +222,136 @@ private void saveThePurchasePriceInOrderProduct(List<OrderProduct> orderProducts
+        Optional<Order> optionalOrder = orderRepository.findById(orderId);
+        Optional<Product> optionalProduct = productRepository.findById(productId);
+        if (optionalOrder.isEmpty()) {
+            throw new OrderNotFoundException("Order with id " + orderId + " not found!");
+        }
+        if (optionalProduct.isEmpty()) {
+            throw new ProductNotFoundException("Product with id " + productId + " not found!");
+        }
+        if (newQuantity < 0) {
+            throw new QuantityNotAvailableException("The quantity cannot be negative number");
+        }
+        Order order = optionalOrder.get();
+        Product product = optionalProduct.get();
+
+        if (!isQuantityAvailable(product, newQuantity)) {
+            throw new QuantityNotAvailableException("The quantity that you are trying to set is not available!");
+        }
+        List<OrderProduct> currentOrderProducts = orderProductRepository.findAllByOrder(order);
+        Long currentOrderProductId;
+        try {
+            currentOrderProductId = orderProductRepository.findByOrderIdAndProductId(orderId, productId).getId();
+        } catch (RuntimeException e) {
+            throw new RuntimeException("Order/Product was not found!");
+        }
+        if (newQuantity == 0) {
+            orderProductRepository.deleteById(currentOrderProductId);
+            return;
+        }
+        for (OrderProduct op : currentOrderProducts) {
+            if (op.getProduct().getId().equals(productId)) {
+                op.setQuantity(newQuantity);
+                orderProductRepository.save(op);
+                return;
+            }
+        }
+    }
+
+    private boolean isQuantityAvailable(Product product, int newQuantity) {
+        return product.getQuantity() >= newQuantity;
+    }
+
+    public ResponseEntity<?> updateQuantity(Long productId, Long orderId, int quantity) {
+        try {
+            updateProductQuantity(orderId, productId, quantity);
+        } catch (RuntimeException e) {
+            return ResponseEntity.badRequest().body(e.getMessage());
+        }
+        Order basketOrder = orderRepository.findById(orderId).get();
+        BigDecimal totalPrice = calculateTotalPrice(getBasketProducts(basketOrder));
+        List<ProductResponseDto> basketProducts = getBasketProducts(basketOrder);
+        return ResponseEntity.ok(new BasketResponse(basketProducts, totalPrice));
+    }
+
+    private String formatDeliveryAddress(User user) {
+        return user.getAddress().getCity().getName() + ", " + user.getAddress().getStreetName();
+    }
+
+    private BigDecimal calculateTotalPriceAndFillProductResponses(List<OrderProduct> products, List<ProductResponse> productResponses) {
+        BigDecimal total = BigDecimal.ZERO;
+        for (OrderProduct op : products) {
+            ProductResponse productResponse = new ProductResponse(op.getProduct().getName(), op.getQuantity());
```

Comment:
Why do we need the productResponse in this method?

Reply:
I removed it. Now separate method will add the products and their quantities to the list of the response dto.